src/aspire/source/xform.py

```diff
@@ -0,0 +1,388 @@
+        self.indices = indices
+        self.n_xforms = len(unique_xforms)
+        self.unique_xforms = unique_xforms
+
+        # A list of references to individual Xform objects, with possibly multiple references pointing to
+        # the same Xform object.
+        self.xforms = [unique_xforms[i] for i in indices]
+
+    def _forward_or_adjoint(self, im, indices, which):
+        """
+        Apply either a forward or adjoint transformations to `im`, depending on the value of the 'which' parameter.
+        :param im: The incoming Image object on which to apply the forward or adjoint transformations.
+        :param indices: The indices of the transformations to apply.
+        :param which: Either 'forward' or 'adjoint'.
+        :return: An Image object as a result of applying forward or adjoint transformation to `im`.
+        """
+        assert which in ('forward', 'adjoint')
+        # Ensure that we will be able to apply all transformers to the image
+        assert self.n_indices >= im.n_images, f'Can process Image object of max depth {self.n_indices}. Got {im.n_images}.'
+
+        im_data = im.asnumpy().copy()
+
+        # For each individual transformation
+        for i in range(self.n_xforms):
+            # Get the indices corresponding to that transformation
+            idx = np.where(self.indices == i)[0]
+            # For the incoming Image object, find out which transformation indices are applicable
+            idx = np.intersect1d(idx, indices)
+            # For the transformation indices we found, find the indices in the Image object that we'll use
+            im_data_indices = np.where(np.isin(indices, idx))[0]
+            # Apply the transformation to the selected indices in the Image object
+            if len(im_data_indices) > 0:
+                fn_handle = getattr(self.unique_xforms[i], which)
+                im_data[:, :, im_data_indices] = fn_handle(Image(im_data[:, :, im_data_indices])).asnumpy()
+        return Image(im_data)
+
+    def _forward(self, im, indices):
+        return self._forward_or_adjoint(im, indices, 'forward')
+
+    def _adjoint(self, im, indices):
+        return self._forward_or_adjoint(im, indices, 'adjoint')
+
+    def downsample(self, resolution):
+        for xform in self.unique_xforms:
+            xform.downsample(resolution)
+        super().downsample(resolution)
+
+
+def _apply_transform(xform, im, indices, adjoint=False):
+    """
+    A simple global function (i.e. not a method) that is capable of being cached by joblib.Memory object's `cache`
+    method.
+    """
+    if not adjoint:
+        logger.info('  Applying ' + str(xform))
+        return xform.forward(im, indices=indices)
+    else:
+        logger.info('  Applying Adjoint ' + str(xform))
+        return xform.adjoint(im, indices=indices)
+
+
+class Pipeline(Xform):
+    """
+    A Pipeline is a Xform made up of individual transformation steps (i.e. multiple Xform objects).
```

Comment:
Make sure docstrings are shorter than 80 characters in line width. Also, backticks around class names.

Reply:
Added backticks. Issue #69 for line width (which we can tackle in one fell swoop)